File: Baekjoon/삼성SW기출/17779.py
n = int(input())
board = [[]]
for _ in range(n):
    board.append([0] + list(map(int, input().split())))

def count(x, y, d1, d2):  # 기준점 (x, y)와 경계의 길이 d1, d2일 때 -> 인구가 가장 많은 선거구와 가장 적은 선거구의 인구 차이
    total = [0] * 5  # 각 선거구의 인구 수
    divide = [[0] * (n + 1) for _ in range(n + 1)]  # 선거구

    # 경계선
    for i in range(d1 + 1):
        divide[x + i][y - i] = 1  # 왼쪽, 위로 향하는 대각선
        divide[x + d2 + i][y + d2 - i] = 1  # 오른쪽, 위로 향하는 대각선
    for j in range(d2 + 1):
        divide[x + j][y + j] = 1  # 오른쪽, 아래로 향하는 대각선
        divide[x + d1 + j][y - d1 + j] = 1  # 왼쪽, 아래로 향하는 대각선

    # 5번 선거구 채우기
    for i in range(x + 1, x + d1 + d2):
        flag = False
        for j in range(n + 1):
            if divide[i][j] == 1:  # 경계구역 다시 만날 때까지
                flag = not flag
            else:
                if flag:
                    divide[i][j] = 1

    # 나눈 구역으로 인구 수 카운트하기
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if i < x + d1 and j <= y and divide[i][j] != 1:
                total[0] += board[i][j]
            elif i <= x + d2 and y < j and divide[i][j] != 1:
                total[1] += board[i][j]
            elif x + d1 <= i and j < y - d1 + d2 and divide[i][j] != 1:
                total[2] += board[i][j]
            elif x + d2 < i and y - d1 + d2 <= j and divide[i][j] != 1:
                total[3] += board[i][j]
            elif divide[i][j] == 1:
                total[4] += board[i][j]

    return max(total) - min(total)

answer = int(1e9)
for x in range(1, n + 1):
    for y in range(1, n + 1):
        for d1 in range(1, n + 1):
            for d2 in range(1, n + 1):
                if 1 <= x < x + d1 + d2 <= n and 1 <= y - d1 < y < y + d2 <= n:
                    answer = min(answer, count(x, y, d1, d2))
print(answer)

File: Baekjoon/삼성SW기출/17779.java
import java.util.*;
import java.io.*;

public class Main {

    static int n;
    static int answer = Integer.MAX_VALUE;
    static int[][] board;
    static int totalCount = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        board = new int[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
                totalCount += board[i][j];
            }
        }

        for (int x = 0; x < n; x++) {
            for (int y = 0; y < n; y++) {
                for (int d1 = 1; d1 < n; d1++) {
                    for (int d2 = 1; d2 < n; d2++) {
                        if (x + d1 + d2 >= n) continue;
                        if (y - d1 < 0 || y + d2 >= n) continue;

                        countRegions(x, y, d1, d2);
                    }
                }
            }
        }

        System.out.println(answer);
    }

    static void countRegions(int x, int y, int d1, int d2) {
        boolean[][] border = new boolean[n][n];
        int[] count = new int[5];

        // 경계선 표시
        for (int i = 0; i <= d1; i++) {
            border[x + i][y - i] = true;
            border[x + d2 + i][y + d2 - i] = true;
        }

        for (int i = 0; i <= d2; i++) {
            border[x + i][y + i] = true;
            border[x + d1 + i][y - d1 + i] = true;
        }

        // 1구역 인구수
        for (int i = 0; i < x + d1; i++) {
            for (int j = 0; j <= y; j++) {
                if (border[i][j]) break;
                count[0] += board[i][j];
            }
        }

        // 2구역 인구수
        for (int i = 0; i <= x + d2; i++) {
            for (int j = n - 1; j > y; j--) {
                if (border[i][j]) break;
                count[1] += board[i][j];
            }
        }

        // 3구역 인구수
        for (int i = x + d1; i < n; i++) {
            for (int j = 0; j < y - d1 + d2; j++) {
                if (border[i][j]) break;
                count[2] += board[i][j];
            }
        }

        // 4구역 인구수
        for (int i = x + d2 + 1; i < n; i++) {
            for (int j = n - 1; j >= y - d1 + d2; j--) {
                if (border[i][j]) break;
                count[3] += board[i][j];
            }
        }

        // 5구역 인구수
        count[4] = totalCount;
        for (int i = 0; i < 4; i++) {
            count[4] -= count[i];
        }

        Arrays.sort(count);
        answer = Math.min(answer, count[4] - count[0]);
    }
}